treewide: replace substituteAll with replaceVars (the end)

wolfgangwalther opened the pull request:
This replaces all remaining uses of `substituteAll` (the nix function) with `replaceVars` and `replaceVarsWith` for #237216.

The commit was created by:
- Running my [ast-grep based replacer](https://github.com/technowledgy/refactor-tractor/blob/main/rules/no-substitute-all-nix.yml) for substituteAll -> replaceVars.
- Fix everything that couldn't be done automatically.

Tested with:
- [`lint-replace-vars`](https://github.com/technowledgy/refactor-tractor/blob/main/rules/no-broken-replace-vars.yml) and
- [`build-replace-vars`](https://github.com/technowledgy/refactor-tractor/blob/main/scripts/build-replace-vars.nu)

The idea behind this approach to testing is described in https://github.com/technowledgy/refactor-tractor?tab=readme-ov-file#build-replace-vars.

## How to review?

Hop on the [refactor tractor](https://github.com/technowledgy/refactor-tractor):
- checkout this branch locally
- clone the refactor-tractor repo and enter its devshell
- run `build-replace-vars <path-to-checked-out-nixpkgs>`
- run `lint-replace-vars <path-to-checked-out-nixpkgs>`

Right now, there will be two build failures for lua, which are being fixed in #376792 and IIRC three linter errors, which have been fixed in #376728, but haven't made their way into this PR, yet.

If you want to see what's happening for the build command, do `NU_LOG_LEVEL=debug build-replace-vars ...` instead and see the expressions we're building fly by.

Instead of reviewing all the changes made in this PR, you could also re-run the `no-substitute-all <nixpkgs-path>` command from that repo on the base commit of this branch - and then compare the result of that to the HEAD of this branch. This would show the manual changes I made on top, a lot less to look at. On the flip-side - you'd probably have to dig into the ast-grep rules a bit as well ;)

## Things done

- Built on platform(s)
  - [x] x86_64-linux (via tools mentioned above)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
treewide: replace substituteAll with replaceVars

philiptaron commented:
> Will probably only rebase, when the first review is through and the approach is approved in principle.

I approve it in principle, and I hope along the way to learn more about how to use ast-grep to do this job. 🙂 

Rather than rebasing, consider splitting the changes to 

1. `pkgs/development/python-modules/psycopg/default.nix`
2. `pkgs/development/python-modules/pypass/default.nix`
3. `pkgs/servers/sql/postgresql/generic.nix`

... out into their own PRs. The postgres ones I think you're the conflicting agent, and pypass looks trivially mergable.

philiptaron commented on nixos/maintainers/scripts/lxd/lxd-container-image.nix:
Tested with `nix-build nixos/release.nix -A lxdContainerImage.x86_64-linux` ✅

philiptaron commented on nixos/maintainers/scripts/lxd/lxd-virtual-machine-image.nix:
Tested with `nix-build nixos/release.nix -A lxdVirtualMachineImage.x86_64-linux` ✅

philiptaron commented on nixos/modules/installer/cd-dvd/iso-image.nix:
Tested with `nix-build nixos/release.nix -A iso_minimal.x86_64-linux` :x:

```
substituteStream() in derivation bpcacbjwamkxnnsai7azykm61v13y9cg-isolinux.cfg-in: ERROR: pattern @bootRoot@ doesn't match anything in file '/nix/store/bpcacbjwamkxnnsai7azykm61v13y9cg-isolinux.cfg-in'
```

philiptaron commented on nixos/modules/services/misc/taskserver/default.nix:
Tested with `nix-build -A nixosTests.taskserver` ✅

philiptaron reviewed (changes requested):
I'm slowly going through the NixOS modules and testing them. One defect that would have been a channel blocker found.

philiptaron reviewed (changes requested):
There are still a lot of broken patches. I didn't de-dupe this list, but all of these failed to build.

```
dropbox-cli
budgie-control-center
cbmc
gdm
gnome-control-center
gnome-initial-setup
gnome-session
gnome-settings-daemon
gnome-shell
gnome-software
gvfs
minigalaxy
muffin
nautilus-python
onboard
pretix
speechd
sway-unwrapped
swayfx-unwrapped
dotnetCorePackages
luajit
gssdp-tools
libpeas2
libvirt
wayqt
webkitgtk_4_0
webkitgtk_4_1
webkitgtk_6_0
clasp-common-lisp
luaPackages.nfd
python3Packages.django_2
python3Packages.django_3
python3Packages.django_4
python3Packages.django_5
python3Packages.glymur
python3Packages.netmap
python3Packages.wandb
python3Packages.wxpython_4_2
kinfocenter
plasma-workspace
udisks
udisks2
home-assistant
networkmanager
```

wolfgangwalther committed:
replaceVars: fail when exemption can't be found

This also changes stdenv's substitute --replace-fail to error out when
the replacement is the same as the search pattern, but can't be found.
This should not cause any problems in existing code, from what I can
tell from grepping nixpkgs.

The exception for pattern==replacement was previously introduced all the
way back in 5ff872aa24983cf3e1cf28bb990042846c1a97ee, but this was
apparently only used to make the check for the warning "simpler".

philiptaron reviewed (approved):
The change looks good to me. I'm doing a big build overnight to get a sense of whether this builds successfully but the approach is now more sound in a couple ways. Thank you @wolfgangwalther!

philiptaron commented:
Build report is good. There are build failures but none due to replaceVars.

vcunat commented:
This broke all ISOs, log e.g.
https://hydra.nixos.org/build/290580097/nixlog/1/tail

vcunat commented:
If someone doesn't come with a fix fast, I'm reverting changes to that file for now.

vcunat commented:
Resolved by https://github.com/NixOS/nixpkgs/pull/383859
